File: services/ReadingService/src/main/java/com/se1853_jv/readingservice/model/ReadingWorkflowNote.java
package com.se1853_jv.readingservice.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.Column;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinColumns;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "reading_workflow_note")
public class ReadingWorkflowNote {

    @EmbeddedId
    private ReadingWorkflowNoteId id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumns({
            @JoinColumn(name = "collection_id", referencedColumnName = "collection_id", insertable = false, updatable = false),
            @JoinColumn(name = "paper_id", referencedColumnName = "paper_id", insertable = false, updatable = false),
            @JoinColumn(name = "user_id", referencedColumnName = "user_id", insertable = false, updatable = false)
    })
    @JsonIgnore
    private ReadingWorkflow readingWorkflow;

    // Helper methods to access composite key fields
    public String getCollectionId() {
        return id != null ? id.getCollectionId() : null;
    }

    public String getPaperId() {
        return id != null ? id.getPaperId() : null;
    }

    public String getUserId() {
        return id != null ? id.getUserId() : null;
    }

    public String getNoteId() {
        return id != null ? id.getNoteId() : null;
    }

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "note_id", referencedColumnName = "id", insertable = false, updatable = false)
    private Note note;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @PrePersist
    public void prePersist() {
        this.createdAt = LocalDateTime.now();
    }
}

File: services/paper-service/src/main/kotlin/com/se1853_jv/model/ReadingWorkflowNote.kt
package com.se1853_jv.model

import jakarta.persistence.*
import java.io.Serializable

/**
 * Junction table: ReadingWorkflow_Note
 * Link Note với ReadingWorkflow (many-to-many relationship)
 */
@Entity
@Table(name = "ReadingWorkflow_Note")
@IdClass(ReadingWorkflowNoteId::class)
data class ReadingWorkflowNote(
    @Id
    @Column(name = "RWCollectionId", length = 36, nullable = false)
    val rwCollectionId: String,

    @Id
    @Column(name = "RWPaperId", length = 36, nullable = false)
    val rwPaperId: String,

    @Id
    @Column(name = "RWUsersid", length = 36, nullable = false)
    val rwUserId: String,

    @Id
    @Column(name = "NoteId", length = 36, nullable = false)
    val noteId: String,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumns(
        JoinColumn(name = "RWCollectionId", referencedColumnName = "CollectionId", insertable = false, updatable = false),
        JoinColumn(name = "RWPaperId", referencedColumnName = "PaperId", insertable = false, updatable = false),
        JoinColumn(name = "RWUsersid", referencedColumnName = "UsersId", insertable = false, updatable = false)
    )
    val readingWorkflow: ReadingWorkflow? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "NoteId", referencedColumnName = "id", insertable = false, updatable = false)
    val note: Note? = null
)

data class ReadingWorkflowNoteId(
    val rwCollectionId: String,
    val rwPaperId: String,
    val rwUserId: String,
    val noteId: String
) : Serializable







